From 1f0989e96a816a980340c93382c49239dd59da04 Mon Sep 17 00:00:00 2001
From: kikairoya <kikairoya@gmail.com>
Date: Wed, 30 Sep 2026 20:56:31 +0900
Subject: [PATCH 38/65] [ORC] Disable __unw_add_dynamic_eh_frame_section on
 Cygwin

Causes a link error:
```
    ld.lld: error: undefined symbol: __unw_add_dynamic_eh_frame_section
    >>> referenced by libLLVMOrcTargetProcess.a(RegisterEHFrames.cpp.o):(llvm::orc::registerEHFrameSection(void const*, unsigned long))
    >>> referenced by libLLVMOrcTargetProcess.a(RegisterEHFrames.cpp.o):(registerEHFrameWrapper(llvm::orc::ExecutorAddrRange))
```
---
 llvm/include/llvm/Support/Compiler.h                      | 5 +++++
 .../Orc/TargetProcess/RegisterEHFrames.cpp                | 8 +++++---
 2 files changed, 10 insertions(+), 3 deletions(-)

diff --git a/llvm/include/llvm/Support/Compiler.h b/llvm/include/llvm/Support/Compiler.h
index 33ea019a87be57d8eaac1668446e424013404e2a..d8a13a261b77a499b051ab708dda607328dbd562 100644
--- a/llvm/include/llvm/Support/Compiler.h
+++ b/llvm/include/llvm/Support/Compiler.h
@@ -309,6 +309,11 @@
 // FIXME: Provide this for PE/COFF targets.
 #if __has_attribute(weak) && !defined(__MINGW32__) && !defined(__CYGWIN__) &&  \
     !defined(_WIN32)
+#define LLVM_HAS_ATTRIBUTE_WEAK 1
+#else
+#define LLVM_HAS_ATTRIBUTE_WEAK 0
+#endif
+#if LLVM_HAS_ATTRIBUTE_WEAK
 #define LLVM_ATTRIBUTE_WEAK __attribute__((__weak__))
 #else
 #define LLVM_ATTRIBUTE_WEAK
diff --git a/llvm/lib/ExecutionEngine/Orc/TargetProcess/RegisterEHFrames.cpp b/llvm/lib/ExecutionEngine/Orc/TargetProcess/RegisterEHFrames.cpp
index b64f7664f80f3fd83554d15a220033b09292547e..01b5f67f8f04daabef18e46db58a72806c93d13e 100644
--- a/llvm/lib/ExecutionEngine/Orc/TargetProcess/RegisterEHFrames.cpp
+++ b/llvm/lib/ExecutionEngine/Orc/TargetProcess/RegisterEHFrames.cpp
@@ -22,7 +22,9 @@ using namespace llvm::orc::shared;
 namespace llvm {
 namespace orc {
 
-#if !defined(_WIN32)
+// Reference __unw_* only if __attribute__((weak)) is available to avoid force
+// pulling them.
+#if LLVM_HAS_ATTRIBUTE_WEAK
 extern "C" void
 __unw_add_dynamic_eh_frame_section(const void *) LLVM_ATTRIBUTE_WEAK;
 extern "C" void
@@ -130,7 +132,7 @@ Error walkLibunwindEHFrameSection(const char *const SectionStart,
 Error registerEHFrameSection(const void *EHFrameSectionAddr,
                              size_t EHFrameSectionSize) {
   // For libunwind after D111863.
-#if !defined(_WIN32)
+#if LLVM_HAS_ATTRIBUTE_WEAK
   if (__unw_add_dynamic_eh_frame_section &&
       __unw_remove_dynamic_eh_frame_section) {
     __unw_add_dynamic_eh_frame_section(EHFrameSectionAddr);
@@ -157,7 +159,7 @@ Error registerEHFrameSection(const void *EHFrameSectionAddr,
 
 Error deregisterEHFrameSection(const void *EHFrameSectionAddr,
                                size_t EHFrameSectionSize) {
-#if !defined(_WIN32)
+#if LLVM_HAS_ATTRIBUTE_WEAK
   if (__unw_add_dynamic_eh_frame_section &&
       __unw_remove_dynamic_eh_frame_section) {
     __unw_remove_dynamic_eh_frame_section(EHFrameSectionAddr);
-- 
2.55.0

